src/slam6d/sift/autopano-sift-c/doc/autopano.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH AUTOPANO 1 "JANUAR 2005" autopano-sift "User Manual"
.SH NAME
autopano \- Generate panorama project from SIFT keypoint files.
.SH SYNOPSIS
.B autopano
.I [options]
.I output.pto
.I keyfile1[.gz]
.I keyfile2[.gz]
.I [keyfile3[.gz] [..]]
.SH DESCRIPTION
Generate panorama project from keypoint data. First, read in two or more SIFT
keypoint files, then correlate the keypoint data and build a list of matches.
This list undergoes some filtering and the best matches are used to create
control point information. The control point information are writting to a
PTO compatible panorama project file. For example, programs such as
.BR hugin
can read it.

Prior to any further use of the PTO, you ABSOLUTELY HAVE TO ADJUST THE FOCAL
LENGTH of all the images. This can be done in hugin, in the "Camera and Lens"
tab.
.SH OPTIONS
.IP "--ransac <on|off|1|0>"
Switch RANSAC postfiltration on or off. Default is on. There are only a few
reasons to switch this off: if the keypoint density and matches are very
sparse, RANSAC may filter too much. Or, if the lens geometry used is unusual
(fish eye, very wide angle, micropanorama, ...) and does not resemble a
rectilinear projection. That is, if you get really bad results with RANSAC on,
disable it.
.IP "--maxmatches <count>"
Set the maximum number of control point pairs you want to generate per image
pair. The default is 16, so for each image pair, there are no more than 16
control points generated. If there are more control points to choose from, an
area maximization metric is applied to keep the matches that cover most of the
image area. You can disable this by setting count to zero. Then all matches
are kept.
.IP "--disable-areafilter"
At the final stage of creating control points, there is a list of matches for
every image pair that overlaps. All this matches are thought to be correct and
have been filtered using the RANSAC algorithm. However, often there are more
matches available than the user wants to keep (see the "--maxmatches" option).
In previous versions we applied a simple metric to pick out the matches that
have a good matching score. Now, we have an area maximization algorithm, that
maximizes the area covered by the matches. In general, this leads to better
coverage of the image overlap area. However, if you want the old behaviour,
that only considers the SIFT matching score of every match, enable this
option.
.IP "--integer-coordinates"
Use natural number coordinates in the PTO file for the found matches.  Default
is to use sub-pixel float coordinates to increase precision.  You may want to
try this option in case you use another frontend than hugin and you experience
problems with the PTO files created by autopano-sift.
.IP "--absolute-pathnames <on|off|1|0>"
Store the absolute pathnames of the image files in the PTO file.  Normally, in
case all images are in the same directory and the output PTO file is also
saved in this directory, only the image filenames are used.  Otherwise the
absolute pathnames are used.  Using this option you can enforce the behaviour.
.IP output.pto
The filename of the PTO project file to generate.
.IP keyfile[.gz]
The SIFT XML keypoint data file, as generated with the
.BR generatekeys (1)
program. If the filename ends in ".gz", transparent
.BR gzip (1)
decompression is used.
.SH ALIGN OPTIONS
.IP "--align"
Enable fully automatic pre-aligning algorithm. This results in yaw, pitch and
rotation coordinates being assigned to the images in the resulting PTO file.
This option is not perfect yet, but in most cases the result is far better
than without using it.

There are a number of conditions on the input images that
can be used with this algorithm. First, the images should all be of the same
dimension, scale and have a simple (rectangular) geometry with roughly
euclidean distances. Second, the order of the input images is considered so
that the first images build an ordered row on the horizon. That is, the first,
second, third, etc. images are strictly aligned left-to-right or right-to-left
and all lie roughly on the horizon. They must also all be of the same rotation
angle, which must be one of three rough cases: bottom-is-bottom,
bottom-is-left, bottom-is-right. Bottom-is-top is forbidden. In case the
bottom is either left or right, we estimate its position based on average
keypoint density (also experimental).

In case the horizontal first row is not properly detected, try to increase the
downscale resolution when creating the keypoint data. This will produce more
keypoints which makes it easier to build the horizont-row. In case it still
does not work, you should not enable this option. Please report bugs or
successes with this option.
.IP "--bottom-is-left"
.IP "--bottom-is-right"
Only usable if --align option is enabled. If your input images are 90 or -90
degree tilted for the first row of horizontal images, you can force the
orientation by telling the program where the bottom (floor) of the images is
located: on the left or right side. If this option is not used, the program
will attempt to automatically figure the orientation based on average keypoint
density in the left and right half of the image.
.IP "--generate-horizon <count>"
Generate a horizon from the first row of aligned horizontal images. This can
only be used if the --align option is enabled. Then, up to count number of
horizon control lines are written into the resulting PTO file at the middle of
the first row images. The lines are optimally spaced and you should use values
such as 2, 6 or 14 to get optimal results (the sum of power of two, starting
with 2, 2 + 4, 2 + 4 + 8).
.SH REFINEMENT OPTIONS
.IP "--refine"
Enable the refinement step.  The refining is done as last step before writing
the PTO output file.  For every matching control point pair a small patch in
the original images is extracted at the original resolution.  The image
patches are matched against each other with the highest possible quality of
matching, usually yielding dozens of keypoints.  As this matches are derived
from the original resolution image, their location is more precise and they
are used to refine the original keypoint using one of the two methods below.
Note that enabling the refinement step makes the total process longer, hence
it is not enabled by default.
.IP "--refine-by-middle"
.IP "--refine-by-mean"
Two methods are available to choose the best point from the high resolution
matches.  Refine by middle searches the match closest to the original
keypoints position and keeps only this match.  Refine by mean builds the
geometric center of all keypoints found in the patch and uses this
coordinates.
.IP "--keep-unrefinable <on|off|1|0>"
In case a good match cannot be refined because it is located to near to the
boundary of the image, the original match is kept by default.  To disable this
behaviour, you can use this option, which throws away matches that cannot be
refined.  However, you might want to increase the number of matches to keep
per image pair using the "--maxmatches" option then.
.SH COMPONENT IDENTIFICATION
The program complains about non-connected components in case there is one or
more images or image groups which have no relationship to the rest of the
images. This means there is no way to jump from an image in one group to
another group by just following control point pairs between any two images.
This component identification is crucial for further optimization: if there
is one or more non-connected components, global optimization based on control
point pairs will be impossible.

To fix this problem, first identify the reason why there are no keypoint
matches between the images in the different components. A common reason is
that the images in one component are very diffuse and only have very few
control points (such as images of the sky or water). If this is the case, you
can try to increase the number of keypoints found in each image by increasing
the downscaleResolution parameter of the
.BR generatekeys (1)
program. Or you can add control point pairs between the images of the
individual components manually, using software such as hugin. Another reason
for different components could be that the images do not belong together to
one panorama.
.SH COMMON PROBLEMS
PTO does not optimize/render in hugin. Please check that you adjusted the
camera lens or focal length parameter for each image. We intentionally set it
to a value that will disallow any operation from within hugin as to force the
user to set this parameter. Without knowing this parameter, any work would be
invalid, thats why we force such strict behaviour, sorry.
.SH BUGS
No bugs known, if you find any, please send a bug report to me. I will try to
fix it.
.SH AUTHOR
Sebastian Nowozin <nowozin at cs dot tu dash berlin dot de>
.SH "SEE ALSO"
.BR autopano-sift (7),
.BR autopano-complete (1),
.BR generatekeys (1),
.BR autopanog (1),
.BR showone (1),
.BR showtwo (1)


src/sift/autopano-sift-c/AutoPano.c

/* autopano-sift, Automatic panorama image creation
 * Copyright (C) 2004 -- Sebastian Nowozin
 *
 * This program is free software released under the GNU General Public
 * License, which is included in this software package (doc/LICENSE).
 */

/* Autopano.cs
 *
 * Keypoint file correlation and hugin panorama file creation utility.
 *
 * (C) Copyright 2004 -- Sebastian Nowozin (nowozin@cs.tu-berlin.de)
 *
 * "The University of British Columbia has applied for a patent on the SIFT
 * algorithm in the United States. Commercial applications of this software
 * may require a license from the University of British Columbia."
 * For more information, see the LICENSE file supplied with the distribution.
 */

#include "AutoPanoSift.h"

void Usage ()
{
    WriteLine ("Autopano: Match Keypoints generated by generatekeys\n");
    WriteLine ("  Version %s\n", PACKAGE_VERSION);
	WriteLine ("usage: autopano.exe [options] output.pto keys1.xml keys2.xml [..]\n");
	WriteLine ("Options");
	WriteLine ("  --ransac <on|off|1|0>   Switch RANSAC filtration on or off (default: on)");
	WriteLine ("  --maxmatches <matches>  Use no more than the given number of matches");
	WriteLine ("                          (default: 16, use zero for unlimited)");
						       
	WriteLine ("  --disable-areafilter    Do not use max-area filtration, which is default.");
	WriteLine ("                          See manpage for details.");
	WriteLine ("  --integer-coordinates   Truncate match coordinates to integer numbers.");
	WriteLine ("  --absolute-pathnames <on|off|1|0>   Use the absolute pathname of the image");
	WriteLine ("                          file in the PTO output file. Disabled by default.");
	WriteLine ("");
	
	WriteLine ("Alignment options");
	WriteLine ("  --align                 Automatically pre-align images in PTO file.");
	WriteLine ("  --bottom-is-left");
	WriteLine ("  --bottom-is-right       Use in case the automatic algorithm fails.");
	WriteLine ("  --generate-horizon <c>  Generate up to 'c' horizon lines.");
	WriteLine ("");
	
	WriteLine ("Refinement options");
	WriteLine ("  --refine                Refine the found control points using the");
	WriteLine ("                          original images.");
	WriteLine ("  --refine-by-middle      Use the best middle point to refine (default).");
	WriteLine ("  --refine-by-mean        Use the mean of the patches control points.");
	WriteLine ("  --keep-unrefinable <on|off|1|0>");
	WriteLine ("                          Keep unrefinable matches (default: on).");
										
	WriteLine ("output.pto: The output PTO panorama project file.");
	WriteLine ("    The filename can be \"-\", then stdout is used");
	WriteLine ("key<n>.xml: The keypoint input files.");
	WriteLine ("    The input files can be gzip compressed, but require the \".gz\" extension\n    then.");
	WriteLine ("");
	WriteLine ("Notice: for the aligning to work, the input images shall be");
	WriteLine ("  1. All of the same dimension and scale");
	WriteLine ("  2. The first images must be an ordered row. See manpage.");
	WriteLine ("");
}

typedef struct Resolution Resolution;
struct Resolution
{
	int x, y;
};

Resolution* Resolution_new0()
{
	Resolution* self = (Resolution*)malloc(sizeof(Resolution));
	return self;
}

Resolution* Resolution_new(int x, int y) 
{
	Resolution* self = Resolution_new0();
	self->x = x;
	self->y = y;
	return self;
}

void Resolution_delete(Resolution* self) 
{
	if (self) {
		free(self);
	}
}

int Resolution_CompareTo (Resolution* self, int x, int y)
{
	
	if (self->x == x && self->y == y)
		return (0);
	
	return (1);
}




// The maximum radius to consider around a keypoint that is refined.  That
// is, at most a patch of a maximum size of twice this value in both
// horizontal and vertical direction is extracted.
int RefinementRadiusMaximum = 96;
int (*refineHandler)(int index, int total);


void RefineKeypoints (ArrayList* msList,
		      bool selectMiddlePoint, bool neverLosePoints);
DisplayImage* ExtractPatch (DisplayImage* large,
			    int px, int py, double scale, int* radius);
ArrayList* ExtractKeypoints (DisplayImage* pic);
bool YesNoOption (char* optionName, char* val);

// selectMiddlePoint: if true, select the middle point in the patch,
//   otherwise build the mean
// neverLosePoints: if true, and if we cannot do the refinement, still use
//   the control point.
void RefineKeypoints (ArrayList* msList,
		bool selectMiddlePoint, bool neverLosePoints)
{
	DisplayImage* pic1 = NULL;
	DisplayImage* pic2 = NULL;
	char* pic1Name = NULL;
	char* pic2Name = NULL;

	/* Keep stats for the refineHandler delegate
	 */
	int totalRefines = 0;
	int doneRefines = 0;
	int i;
	for(i=0; i<ArrayList_Count(msList); i++) {
		MatchSet* ms = (MatchSet*) ArrayList_GetItem(msList, i);
		int j;
		for(j=0; j<ArrayList_Count(ms->matches); j++) {
			ArrayList_GetItem(ms->matches, j);
			totalRefines += 1;
		}
	}

 
	for(i=0; i<ArrayList_Count(msList); i++) {
		MatchSet* ms = (MatchSet*) ArrayList_GetItem(msList, i);
		WriteLine ("  between \"%s\" and \"%s\"",
			   ms->file1, ms->file2);
		
		if (pic1Name != ms->file1) {
			pic1Name = ms->file1;
			pic1 = DisplayImage_new (ms->file1);
		}
		if (pic2Name != ms->file2) {
			pic2Name = ms->file2;
			pic2 = DisplayImage_new (ms->file2);
		}
		/*WriteLine ("pair: %s, %s, %d keypoint matches",
		  ms->file1, ms->file2, ArrayList_Count(ms->Matches));*/
		
		ArrayList* refinedMatches = ArrayList_new0 (NULL);

		int j;
		for(j=0; j<ArrayList_Count(ms->matches); j++) {
			Match* m = (Match*) ArrayList_GetItem(ms->matches, j);

			int p1x = (int) (m->kp1->x + 0.5);
			int p1y = (int) (m->kp1->y + 0.5);
			int p1radius;
			DisplayImage* patch1 = ExtractPatch (pic1, p1x, p1y,
							    m->kp1->scale, &p1radius);

			int p2x = (int) (m->kp2->x + 0.5);
			int p2y = (int) (m->kp2->y + 0.5);
			int p2radius;
			DisplayImage* patch2 = ExtractPatch (pic2, p2x, p2y,
							     m->kp2->scale, &p2radius);

			/* Call the refine handler delegate in case there is one to
			 * inform the callee of a single refining step (for progress
			 * bar displays and such).
			 */
			doneRefines += 1;
			if (refineHandler != NULL)
				refineHandler (doneRefines, totalRefines);

			// Skip over keypoint matches we cannot refine as part of the
			// image lies outside.
			if (patch1 == NULL || patch2 == NULL) {
				if (neverLosePoints)
					ArrayList_AddItem(refinedMatches, m);
                                DisplayImage_delete(patch1);
                                DisplayImage_delete(patch2);
				continue;
			}

			// Otherwise, run the SIFT algorithm on both small patches.
			ArrayList* p1kp = ExtractKeypoints (patch1);
			ArrayList* p2kp = ExtractKeypoints (patch2);
			/*WriteLine ("p1kp = %d, p2kp = %d", ArrayList_Count(p1kp),
			  ArrayList_Count(p2kp));*/
			
			// Apply the matching, RANSAC enabled.
			MultiMatch* mm = MultiMatch_new0 ();
			mm->verbose = false;

			ArrayList* matches = NULL;
			matches = MultiMatch_TwoPatchMatch (mm, p1kp,
							    patch1->width, patch1->height, p2kp, patch2->width,
							    patch2->height, true);
                        DisplayImage_delete(patch1);
                        DisplayImage_delete(patch2);

			/* In case there are less than three keypoints in the
			 * two patches, we ignore them all.
			 */
			if (0 /*was exception ???*/ ) { 
				matches = NULL;
			}

			if (matches == NULL || ArrayList_Count(matches) != 1) {
				if (neverLosePoints)
					ArrayList_AddItem(refinedMatches, m);

				continue;
			}

			MatchSet* pSet = (MatchSet*) ArrayList_GetItem(matches, 0);

			// Now get the real new control point coordinates from the
			// patches.  We have two options and assume all points are
			// equal quality-wise:
			//   a) Select the one that is most in the middle
			//      (selectMiddlePoint == true)
			//   b) Build the mean of all the control point matches in the
			//      patches (selectMiddlePoint == false).
			double kp1X = 0.0;
			double kp1Y = 0.0;
			double kp2X = 0.0;
			double kp2Y = 0.0;
			double kpMidDist = Double_PositiveInfinity;

			int k;
			for(k=0; k<ArrayList_Count(pSet->matches); k++) {
				Match* pM = (Match*) ArrayList_GetItem(pSet->matches, k);
				if (selectMiddlePoint) {
					double dist = sqrt (
						pow (pM->kp1->x - p1radius, 2.0) +
						pow (pM->kp1->y - p1radius, 2.0));
					
					if (dist < kpMidDist) {
						kpMidDist = dist;
						
						kp1X = pM->kp1->x;
						kp1Y = pM->kp1->y;
						
						kp2X = pM->kp2->x;
						kp2Y = pM->kp2->y;
					}
				} else {
					kp1X += pM->kp1->x;
					kp1Y += pM->kp1->y;
					
					kp2X += pM->kp2->x;
					kp2Y += pM->kp2->y;
				}

				/*WriteLine ("(%g, %g) matches (%g, %g)",
				  pM->kp1->x, pM->kp1->y, pM->kp2->x, pM->kp2->y);*/
			}

			if (selectMiddlePoint == false) {
				kp1X /= (double) ArrayList_Count(pSet->matches);
				kp1Y /= (double) ArrayList_Count(pSet->matches);
				kp2X /= (double) ArrayList_Count(pSet->matches);
				kp2Y /= (double) ArrayList_Count(pSet->matches);
			}

			kp1X += p1x - p1radius;
			kp1Y += p1y - p1radius;

			kp2X += p2x - p2radius;
			kp2Y += p2y - p2radius;

			Match* mn = Match_clone (m);

			// Adjust the original keypoints location to be the mean of
			// all the highly precise superresolution points.
			mn->kp1->x = kp1X;
			mn->kp1->y = kp1Y;

			mn->kp2->x = kp2X;
			mn->kp2->y = kp2Y;

			/*WriteLine ("MASTER POINT MATCH: (%g,%g) to (%g,%g)",
			  kp1X, kp1Y, kp2X, kp2Y);*/

			ArrayList_AddItem(refinedMatches, mn);
			/*
			  DisplayImage_Save (patch1, "patch-1.jpg");
			  DisplayImage_Save (patch2, "patch-2.jpg");
			  exit (0);
			*/
		}

		ms->matches = refinedMatches;
	}
}

/** Extract a small image patch from a larger image, centered at the given
 * coordinates.
 */
DisplayImage* ExtractPatch (DisplayImage* large,
			    int px, int py, double scale, int* radius)
{
	*radius = (int) (9.0 * scale + 0.5);
	if (*radius > RefinementRadiusMaximum)
		*radius = RefinementRadiusMaximum;

	/*WriteLine ("patch centered at (%d,%d), scale %g, radius = %d",
	  px, py, scale, *radius);*/

	int pxe = px + *radius;
	int pye = py + *radius;
	px -= *radius;
	py -= *radius;

	if (px < 0 || py < 0 || pxe >= large->width || pye >= large->height) {
		/*WriteLine ("   (%d,%d)-(%d,%d) out of (0,0)-(%d,%d)",
		  px, py, pxe, pye, large->width, large->height);*/

		return (NULL);
	} else {
		//WriteLine ("   extracting patch");
	}
	DisplayImage* patch = DisplayImage_Carve (large, px, py, *radius*2, *radius*2);

	return (patch);
}

/** Produce keypoints for a small image patch.
 */
ArrayList* ExtractKeypoints (DisplayImage* pic)
{
	ImageMap* picMap = DisplayImage_ConvertToImageMap (pic);
	
	LoweFeatureDetector* lf = LoweFeatureDetector_new0 ();
	LoweFeatureDetector_SetPrintWarning(false);
	LoweFeatureDetector_SetVerbose(false);
	LoweFeatureDetector_DetectFeatures (lf, picMap);
	
        ArrayList* res = LoweFeatureDetector_GlobalNaturalKeypoints(lf);
        lf->globalNaturalKeypoints = NULL; // Make sure res won't get deleted.
        LoweFeatureDetector_delete(lf);
        return res;
}

bool YesNoOption (char* optionName, char* val)
{
    if (strcmp (val, "1") == 0 || strcmp (val, "on") == 0)
	return (true);
    else if (strcmp (val, "0") == 0 || strcmp (val, "off") == 0)
	return (false);
    
    FatalError ("'%s' is no valid truth value for option '%s'. Please see manpage for help.",
		val, optionName);
    return false;
}

void WritePTOFile (FILE* pto, MultiMatch* mm,
		   ArrayList* msList, BondBall* bb, int generateHorizon, bool integerCoordinates,
		   bool useAbsolutePathnames)
{
	fprintf(pto, "# Hugin project file\n");
	fprintf(pto, "# automatically generated by autopano-sift, available at\n");
	fprintf(pto, "# http://cs.tu-berlin.de/~nowozin/autopano-sift/\n\n");
	fprintf(pto, "p f2 w3000 h1500 v360  n\"JPEG q90\"\n");
	fprintf(pto, "m g1 i0\n\n");
	
	int imageIndex = 0;
	HashTable* imageNameTab = HashTable_new0 (NULL, NULL);
	ArrayList* resolutions = ArrayList_new0 (Resolution_delete);
	int i;
	for(i=0; i<ArrayList_Count(mm->keySets); i++) {
		KeypointXMLList* kx = (KeypointXMLList*) ArrayList_GetItem(mm->keySets, i);
		HashTable_AddItem(imageNameTab, kx->imageFile, (void*)imageIndex);
		ArrayList_AddItem(resolutions, Resolution_new (kx->xDim, kx->yDim));
		
		char*  imageFile = kx->imageFile;
		
		// If the resolution was already there, use the first image with
		// the exact same resolution as reference for camera-related
		// values.
		
		int refIdx;
		for (refIdx = 0 ; refIdx < (ArrayList_Count(resolutions) - 1) ; ++refIdx) {
			if (Resolution_CompareTo((Resolution*) ArrayList_GetItem(resolutions, refIdx), kx->xDim, kx->yDim) == 0)
				break;
		}
		if (refIdx == (ArrayList_Count(resolutions) - 1))
			refIdx = -1;
		
		Position* pos = bb == NULL ? NULL :
			(Position*) HashTable_GetItem(bb->positions, imageFile);
		/*
		  if (pos != NULL) {
		  WriteLine ("yaw %g, pitch %g, rotation %g",
		  pos->yaw, pos->pitch, pos->rotation);
		  }*/

		double yaw = 0.0, pitch = 0.0, rotation = 0.0;
		if (pos != NULL) {
			yaw = pos->yaw;
			pitch = pos->pitch;
			rotation = pos->rotation;
		}
		
		if (imageIndex == 0 || refIdx == -1) {
			fprintf(pto, "i w%d h%d f0 a0 b-0.01 c0 d0 e0 p%g r%g v180 y%g  u10 n\"%s\"\n",
				kx->xDim, kx->yDim, pitch, rotation, yaw, imageFile);
		} else {
			fprintf(pto, "i w%d h%d f0 a=%d b=%d c=%d d0 e0 p%g r%g v=%d y%g  u10 n\"%s\"\n",
				kx->xDim, kx->yDim, refIdx, refIdx, refIdx, pitch, rotation, refIdx, yaw, imageFile);
		}
		imageIndex += 1;
	}
	
	fprintf(pto, "\nv p1 r1 y1\n\n");
	
	fprintf(pto, "# match list automatically generated\n");
	int j;
	for(j=0; j<ArrayList_Count(msList); j++) {
		MatchSet* ms = (MatchSet*) ArrayList_GetItem(msList, j);
		int k;
		for(k=0; k<ArrayList_Count(ms->matches); k++) {
			Match* m = (Match*) ArrayList_GetItem(ms->matches, k);
			if (integerCoordinates == false) {
				fprintf(pto, "c n%d N%d x%.6f y%.6f X%.6f Y%.6f t0\n",
					(int)HashTable_GetItem(imageNameTab, ms->file1), (int)HashTable_GetItem(imageNameTab, ms->file2),
					m->kp1->x, m->kp1->y, m->kp2->x, m->kp2->y);
			} else {
				fprintf(pto, "c n%d N%d x%d y%d X%d Y%d t0\n",
					(int)HashTable_GetItem(imageNameTab, ms->file1), (int)HashTable_GetItem(imageNameTab, ms->file2),
					(int) (m->kp1->x + 0.5), (int) (m->kp1->y + 0.5),
					(int) (m->kp2->x + 0.5), (int) (m->kp2->y + 0.5));
			}
		}
	}
	
	// Generate horizon if we should
	if (bb != NULL && generateHorizon > 0) {
		WriteLine ("Creating horizon...");
		
		int kMain = 2;
		int hPoints = generateHorizon;
		int horizonPointsMade = 0;

		bool hasGood = true;
		while (hPoints > 0 && hasGood) {
			hasGood = false;
			int kStep = 2 * kMain;

			int p;
			for (p = 0 ; hPoints > 0 && p < kMain ; ++p) {
				double stepSize = ((double) ArrayList_Count(bb->firstRow)) / ((double) kStep);
				double beginIndex = p * stepSize;
				double endIndex = (((double) ArrayList_Count(bb->firstRow)) / (double) kMain) +
					p * stepSize;

// Round to next integer and check if their image distance
// is larger than 1. If its not, we skip over this useless
// horizon point.
				int bi = (int) (beginIndex + 0.5);
				int ei = (int) (endIndex + 0.5);
				if ((ei - bi) <= 1)
					continue;

				hasGood = true;

				bi %= ArrayList_Count(bb->firstRow);
				ei %= ArrayList_Count(bb->firstRow);
				fprintf(pto, "c n%s N%s x%d y%d X%d Y%d t2\n",
					(char*)HashTable_GetItem(imageNameTab, ArrayList_GetItem(bb->firstRow, bi)),
					(char*)HashTable_GetItem(imageNameTab, ArrayList_GetItem(bb->firstRow, ei)),
					((Resolution*) ArrayList_GetItem(resolutions,bi))->x / 2,
					((Resolution*) ArrayList_GetItem(resolutions,bi))->y / 2,
					((Resolution*) ArrayList_GetItem(resolutions,ei))->x / 2,
					((Resolution*) ArrayList_GetItem(resolutions,ei))->y / 2);

				horizonPointsMade += 1;
				hPoints -= 1;
			}

// Increase density for next generation lines
			kMain *= 2;
		}
		WriteLine ("  made %d horizon lines.\n", horizonPointsMade);
	}

	fprintf(pto, "\n# :-)\n\n");
	
	WriteLine ("\nYou can now load the output file into hugin.");
	WriteLine ("Notice: You absolutely must adjust the field-of-view value for the images");

        ArrayList_delete(resolutions);
        HashTable_delete(imageNameTab);
}


int main (int argc, char* argv[])
{
	WriteLine ("autopano-sift, Automatic panorama generation program\n");
  
	if (argc+1 < 3) {
		Usage ();
		exit (1);
	}

	// output to stdout flag
	bool streamout = false;

	// Automatic pre-aligning of images
	bool preAlign = false;
	int bottomDefault = -1;
	int generateHorizon = 0;
  
	// Use RANSAC algorithm match filtration.
	bool useRansac = true;
  
	// Use area based weighting for final match selection.
	bool useAreaFiltration = true;
  
	// Truncate match coordinates to integer numbers.
	bool useIntegerCoordinates = false;
  
	// Use the absolute pathname of the image files in the output PTO
	// file.
	bool useAbsolutePathnames = false;
  
	// Use "keep-best" filtration, keep the maxMatches best.
	int maxMatches = 16;	// default: 16
  
	// Refinement options
	bool refine = false;
	bool refineMiddle = true;
	bool keepUnrefinable = true;
  
	int optionCount = 0;
	int optionN = 1;
	while (optionN < argc &&
	       strlen(argv[optionN]) >= 2 
		   && argv[optionN][0] == '-'
		   && argv[optionN][1] == '-')
		{
			char* optionStr = argv[optionN];

			if (strcmp (optionStr, "--ransac") == 0) {
				useRansac = YesNoOption ("--ransac", argv[optionN + 1]);
				optionN += 2;
			} else if (strcmp (optionStr, "--maxmatches") == 0) {
				if (sscanf(argv[optionN + 1], "%d",  &maxMatches) != 1) {
					WriteLine ("Parameter to maxmatches option invalid. See the usage help.");
					exit (1);
				}
				if (maxMatches < 0) {
					WriteLine ("Maximum number of matches must be positive or zero (unlimited).");
					exit (1);
				}
				optionN += 2;
			} else if (strcmp (optionStr, "--disable-areafilter") == 0) {
				useAreaFiltration = false;
				optionN += 1;
			} else if (strcmp (optionStr, "--integer-coordinates") == 0) {
				useIntegerCoordinates = true;
				optionN += 1;
			} else if (strcmp (optionStr, "--absolute-pathnames") == 0) {
				useAbsolutePathnames = YesNoOption ("--absolute-pathnames", argv[optionN + 1]);
				optionN += 2;
			} else if (strcmp (optionStr, "--align") == 0) {
				preAlign = true;
				optionN += 1;
			} else if (strcmp (optionStr, "--bottom-is-left") == 0) {
				bottomDefault = 0;
				optionN += 1;
			} else if (strcmp (optionStr, "--bottom-is-right") == 0) {
				bottomDefault = 1;
				optionN += 1;
			} else if (strcmp (optionStr, "--generate-horizon") == 0) {
				if (sscanf(argv[optionN + 1], "%d", &generateHorizon) != 1) {
					WriteLine ("Parameter to generate-horizon option invalid. See the usage help.");
					exit (1);
				}
				if (generateHorizon < 0) {
					WriteLine ("The number of horizon lines to generate must be positive.");
					exit (1);
				}

				optionN += 2;
			} else if (strcmp (optionStr, "--refine") == 0) {
				refine = true;
				optionN += 1;
			} else if (strcmp (optionStr, "--refine-by-middle") == 0) {
				refineMiddle = true;
				optionN += 1;
			} else if (strcmp (optionStr, "--refine-by-mean") == 0) {
				refineMiddle = false;
				optionN += 1;
			} else if (strcmp (optionStr, "--keep-unrefinable") == 0) {
				keepUnrefinable = YesNoOption ("--keep-unrefinable", argv[optionN + 1]);
				optionN += 2;
			} else {
				WriteLine ("Usage error. Run \"autopano.exe\" without arguments for help.");
				exit (1);
			}
		}
	optionCount = optionN;
	// is there an output name and at least one input name?
	if( argc - optionN < 2 ){
		WriteLine ("Error. Output name and at least one image name required.");
		Usage();
		exit(1);
	}
	// next arg is either output file name or "-" for stdout
	// anything else beginning with '-' is an error
	if( argv[optionCount][0] == '-' ){
		if( strcmp( argv[optionCount], "-" ) == 0 )streamout = true;
		else {
			WriteLine ("Option error. Run without arguments for help.");
			exit (1);
		}
	}
	
	if (bottomDefault != -1 && preAlign == false) {
		WriteLine ("Please enable automatic alignment (\"--align\") before using the");
		WriteLine ("--bottom-is-* options. Thank you. Run \"autopano.exe\" without");
		WriteLine ("arguments for usage help.");

		exit (1);
	}

	if (generateHorizon > 0 && preAlign == false) {
		WriteLine ("Please enable automatic alignment (\"--align\") before using the");
		WriteLine ("--generate-horizon option. Thank you. Run \"autopano.exe\" without");
		WriteLine ("arguments for usage help.");

		exit (1);
	}

	MultiMatch* mm = MultiMatch_new0 ();
	ArrayList* keyfiles = ArrayList_new0(NULL);
	int i;
	for( i=0; i<argc - 1 - optionCount; i++) {
		ArrayList_AddItem(keyfiles, argv[i+optionCount+1]);
	}

	WriteLine ("Loading keyfiles");
	MultiMatch_LoadKeysets (mm, keyfiles);

	WriteLine ("\nMatching...%s", useRansac == true ? " RANSAC enabled" : "");
	ArrayList* msList = MultiMatch_LocateMatchSets (mm, 3, maxMatches,
							useRansac, useAreaFiltration);

	// Connected component check
	WriteLine ("\nConnected component check...");
	ArrayList* components = MultiMatch_ComponentCheck (mm, msList);
	WriteLine ("Connected component identification resulted in %d component%s:",
		   ArrayList_Count(components), ArrayList_Count(components) > 1 ? "s" : "");

	int compN = 1;
	int j;
	for(j=0; j<ArrayList_Count(components); j++) {
		Component* comp = (Component*) ArrayList_GetItem(components, j);
                char* compstr = Component_ToString(comp);
                WriteLine ("component %d: %s", compN++, compstr);
                free(compstr);
	}

	if (ArrayList_Count(components) > 1) {
		WriteLine ("");
		WriteLine ("Warning: There is one or more components that are not connected through control");
		WriteLine ("         points. An optimization of the resulting PTO will not be possible");
		WriteLine ("         without prior adding of control points between the components listed");
		WriteLine ("         above. Please see the manual page for autopano(1) for details.");
		WriteLine ("");
	} else
		WriteLine ("");

	// BondBall algorithm
	BondBall* bb = NULL;
	if (preAlign) {
		bb = MultiMatch_BuildBondBall (mm, msList, bottomDefault);

		if (bb == NULL) {
			WriteLine ("WARNING: Failed to build bondball as requested. No pre-aligning of images");
			WriteLine ("         takes place.\n");
		}
	}

	if (refine) {
		WriteLine ("Refining keypoints");
		RefineKeypoints (msList, refineMiddle, keepUnrefinable);
	}

	FILE* pto;
	if ( streamout ) {
		pto = stdout;
	} else {
		WriteLine ("Creating output file \"%s\"", argv[optionCount]);
		pto = fopen(argv[optionCount], "w");
	}

	WritePTOFile (pto, mm, msList, bb, generateHorizon, useIntegerCoordinates,
		      useAbsolutePathnames);

	ArrayList_delete(keyfiles);
	ArrayList_delete(components);
	BondBall_delete(bb);
	MultiMatch_delete(mm);
	if ( !streamout )
		fclose(pto);
	return 0;
}

